### PR TITLE
Use `PatientPrefillInfo` for prefilling patient entry information

### PR DESCRIPTION
**Pull request by janhavisinghh**
https://app.clubhouse.io/simpledotorg/story/3215/use-indianhidinfo-for-prefilling-patient-entry-information\

Currently you will see that the nhid number is being shown where the bp passport number is shown. The handling of this is being done in [another story](https://app.clubhouse.io/simpledotorg/story/3535/add-support-for-nhid-in-patiententryscreen)

### CODE REVIEW

**Comment by vinaysshenoy** [comment]
@janhavisinghh This still seems to be a `work in progress` PR. Is this intentional?

**Comment by janhavisinghh** [comment]
@vinaysshenoy yeah, working on to add few more things in this PR which we figured out later. This should be up for review today. ([slack message](https://simpledotorg.slack.com/archives/C01T2MYJB97/p1620835412007700))

**Comment by janhavisinghh** [comment]
@vinaysshenoy this is ready for review. There is on small thing while rebasing I messed up a bit while merging the branches so I've made a [commit](https://github.com/simpledotorg/simple-android/pull/2611/commits/bd403d9db57741302aca96da99635a2c96550927) to fix that. Please, let me know if there is a better way to handle this. Other than that everything should be good to go.

File: app/src/main/java/org/simple/clinic/util/moshi/LocalDateMoshiAdapter.kt [hunk 1]
```diff
@@ -4,22 +4,34 @@ import com.squareup.moshi.FromJson
...
   }
 
   @ToJson
   fun fromLocalDate(date: LocalDate?): String? {
-    return date?.format(formatter)
+    return try {
+      date?.format(formatter)
+    } catch (e: DateTimeParseException) {
+      date?.format(indiaNHIDFormatter)
+    }
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/util/moshi/LocalDateMoshiAdapter.kt:35]: This is the wrong way to support alternate adapters for a specific type based on the source. The right way to do this is to use Moshi's `@JsonQualifier` annotation.

Please look at the relevant documentation here: https://github.com/square/moshi#alternate-type-adapters-with-jsonqualifier

File: app/src/main/java/org/simple/clinic/util/moshi/LocalDateMoshiAdapter.kt [hunk 2]
```diff
@@ -4,22 +4,34 @@ import com.squareup.moshi.FromJson
...
 import java.time.LocalDate
 import java.time.format.DateTimeFormatter
+import java.time.format.DateTimeParseException
+import java.util.Date
 
 class LocalDateMoshiAdapter {
 
   companion object {
     private val formatter = DateTimeFormatter.ISO_DATE!!
+    private val indiaNHIDFormatter = DateTimeFormatter.ofPattern("d/M/yyyy")
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/util/moshi/LocalDateMoshiAdapter.kt:14]: When defining this date formatter, we also should be explicitly setting the locale to `Locale.ENGLISH` here. Otherwise, we might run into issues if users have the device locale set to something other than non-english.

File: app/src/main/java/org/simple/clinic/patient/OngoingNewPatientEntry.kt [hunk 3]
```diff
@@ -118,7 +119,7 @@ data class OngoingNewPatientEntry(
       copy(
           personalDetails = PersonalDetails(
               fullName = patientProfileInfo.fullName,
-              dateOfBirth = null,
+              dateOfBirth = patientProfileInfo.dateOfBirth.format(DateTimeFormatter.ofPattern("dd/MM/yyyy")),
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/patient/OngoingNewPatientEntry.kt:122]: Please define date formatters in `DateFormatterModule` and pass them to the method as a parameter instead of hardcoding them here.

In fact, we already have a formatter for this specific format already defined.

File: app/src/sharedTest/java/org/simple/clinic/TestData.kt [hunk 4]
```diff
@@ -1226,7 +1227,7 @@ object TestData {
       gender: String = "M",
       state: String = "Maharashtra",
       district: String = "Thane",
-      dateOfBirth: String = "12/12/1997",
+      dateOfBirth: LocalDate = LocalDate.parse("12/12/1997", DateTimeFormatter.ofPattern("d/M/yyyy")),
```
Comment (vinaysshenoy) [inline app/src/sharedTest/java/org/simple/clinic/TestData.kt:1231]: What's the need to create a formatter here? Using `LocalDate.parse` should be sufficient, right?

File: app/src/test/java/org/simple/clinic/instantsearch/InstantSearchUpdateTest.kt [hunk 5]
```diff
@@ -315,14 +316,53 @@ class InstantSearchUpdateTest {
...
+  }
+
+  @Test
+  fun `when register new patient is clicked and patient prefill info is not empty and dob is in wrong format, then save it in ongoing patient entry with dob in correct format`(){
+    val facility = TestData.facility()
+    val searchQueryModel = defaultModel
+        .facilityLoaded(facility)
+        .searchQueryChanged("Pat")
+
+    val dateOfBirthWithWrongFormat = LocalDate.parse("2/3/2012", DateTimeFormatter.ofPattern("d/M/yyyy"))
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/instantsearch/InstantSearchUpdateTest.kt:351]: I am not understanding how this is in the wrong format? If we expect the date in the NHID payload to follow a certain pattern, then yes, that can be wrong.

However, here we are creating a valid `LocalDate` and it can never be in the wrong format at all.

Comment (vinaysshenoy) [reply app/src/test/java/org/simple/clinic/instantsearch/InstantSearchUpdateTest.kt:351]: Here, what you are actually testing is whether `LocalDate` is using the `DateTimeFormatter` passed to the constructor which we should not be testing since we don't own it.

File: app/src/main/java/org/simple/clinic/patient/Gender.kt [hunk 6]
```diff
@@ -65,6 +65,14 @@ val Gender.displayLetterRes: Int
       is Unknown -> R.string.gender_unknown_letter
     }
 
+val Gender.displayChar : String
+  get() = when(this) {
+    Male -> "M"
+    Female -> "F"
+    Transgender -> "T"
+    is Unknown -> "U"
+  }
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/patient/Gender.kt:74]: We already have a `displayLetterRes` property which handles different languages. Why are we introducing a hard-coded english character string here?

File: app/src/main/java/org/simple/clinic/scanid/IndiaNHIDGender.kt [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+val IndiaNHIDGender.toGender: Gender
+  get() = when (this) {
+    IndiaNHIDGender.Male -> Gender.Male
+    IndiaNHIDGender.Female -> Gender.Female
+    else -> Gender.Unknown(this.toChar)
+  }
+
+val IndiaNHIDGender.toChar: String
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/scanid/IndiaNHIDGender.kt:49]: This should be `private` since we don't want it to be used anywhere else.

File: app/src/main/java/org/simple/clinic/scanid/IndiaNHIDGender.kt [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  get() = when (this) {
+    IndiaNHIDGender.Male -> "M"
+    IndiaNHIDGender.Female -> "F"
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/scanid/IndiaNHIDGender.kt:52]: Since we technically don't need to handle `Male` and `Female` gender letters, we can just remove this property and use `U` directly in the other property `when` clause, right?

File: app/src/main/java/org/simple/clinic/scanid/IndiaNHIDGender.kt [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    @FromJson
+    fun fromJson(value: String?): IndiaNHIDGender? = TypeAdapter.toEnum(value)
+
+    @ToJson
+    fun toJson(answer: IndiaNHIDGender): String? = TypeAdapter.fromEnum(answer)
+  }
+}
+
+val IndiaNHIDGender.toGender: Gender
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/scanid/IndiaNHIDGender.kt:42]: The `toGender` name implies that this is an action that can be performed, but this has been defined as a property instead of a method. Please either rename the property or convert it to a method so it makes sense.

Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/scanid/IndiaNHIDGender.kt:42]: An alternate suggestion would be to add a static factory method `fromIndiaNHIDGender` to the `Gender` type.

File: app/src/main/java/org/simple/clinic/scanid/IndiaNHIDGender.kt [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+import android.os.Parcelable
+import androidx.annotation.VisibleForTesting
+import com.squareup.moshi.FromJson
+import com.squareup.moshi.ToJson
+import kotlinx.parcelize.Parcelize
+import org.simple.clinic.patient.Gender
+import org.simple.clinic.util.room.SafeEnumTypeAdapter
+
+sealed class IndiaNHIDGender : Parcelable {
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/scanid/IndiaNHIDGender.kt:11]: This does not need to be a sealed class with a `SafeEnumTypeAdapter` since this is only necessary for types which are both present in the API *and* are saved in the local database. Since we do not store the `IndiaNHIDPayload` in the database and only consume it at runtime, we can use a regular enum type.

File: app/src/main/java/org/simple/clinic/newentry/PatientEntryEffectHandler.kt [hunk 11]
```diff
@@ -105,7 +105,13 @@ class PatientEntryEffectHandler @AssistedInject constructor(
           .flatMapSingle { getPatientEntryAndFacility }
           .subscribeOn(scheduler)
           .map { (entry, facility) ->
-            entry.takeIf { it.address != null } ?: entry.withAddress(Address.withDistrictAndState(facility.district, facility.state))
+            if (entry.address != null && entry.address.doesNotHaveDistrictAndState) {
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/newentry/PatientEntryEffectHandler.kt:108]: This change is unrelated to this PR, please extract it to a different PR.

**Review by vinaysshenoy — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/simple/clinic/scanid/ScanSimpleIdScreen.kt [hunk 12]
```diff
@@ -279,9 +279,13 @@ class ScanSimpleIdScreen : BaseScreen<
...
-  override fun openPatientSearch(additionalIdentifier: Identifier?, initialSearchQuery: String?) {
-    val keyToPush = InstantSearchScreenKey(additionalIdentifier = additionalIdentifier,
-        initialSearchQuery = null)
+  override fun openPatientSearch(additionalIdentifier: Identifier?, initialSearchQuery: String?, patientPrefillInfo: PatientPrefillInfo?) {
+    val keyToPush = if (features.isEnabled(Feature.InstantSearch)) {
+      InstantSearchScreenKey(additionalIdentifier = additionalIdentifier,
+          initialSearchQuery = null)
+    } else {
+      PatientSearchScreenKey(additionalIdentifier).wrap()
+    }
```
[the remark this answers is not in the document]
Comment (janhavisinghh) [reply app/src/main/java/org/simple/clinic/scanid/ScanSimpleIdScreen.kt]: Ok, so this is where I messed up the rebase. I am fixing this. This commit only adds patientPrefillInfo as a parameter in the openPatientSearch method

File: app/src/test/java/org/simple/clinic/scanid/ScanSimpleIdUpdateTest.kt [hunk 13]
```diff
@@ -126,10 +126,32 @@ class ScanSimpleIdUpdateTest {
         .whenEvent(PatientSearchByIdentifierCompleted(patients, identifier))
         .then(assertThatNext(
             hasModel(defaultModel.notSearching()),
-            hasEffects(OpenPatientSearch(identifier, null))
+            hasEffects(OpenPatientSearch(identifier, null, null))
         ))
   }
 
+  @Test
+  fun `when national health identifier is scanned and patient is not found, then send the identifier to parent screen with patient prefill info`() {
```
[the remark this answers is not in the document]
Comment (janhavisinghh) [reply app/src/test/java/org/simple/clinic/scanid/ScanSimpleIdUpdateTest.kt:134]: it is referring to patient search screen. I will make it more explicit.

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt [hunk 14]
```diff
@@ -27,13 +29,17 @@ data class InstantSearchModel(
   val isAdditionalIdentifierAnNHID: Boolean
     get() = additionalIdentifier?.type == IndiaNationalHealthId && hasAdditionalIdentifier
 
+  val hasPatientPrefillInfo: Boolean
```
[the remark this answers is not in the document]
Comment (janhavisinghh) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt:32]: `hasPatientPrefillInfoAndAdditionalIdentifierIsNHID` this sounds good?

File: app/src/main/java/org/simple/clinic/home/patients/PatientsTabScreen.kt [hunk 15]
```diff
@@ -212,14 +212,10 @@ class PatientsTabScreen : BaseScreen<
       .map { SimpleVideoClicked }
 
   override fun openPatientSearchScreen(additionalIdentifier: Identifier?) {
-    val screenKey = if (features.isEnabled(Feature.InstantSearch)) {
```
[the remark this answers is not in the document]
Comment (janhavisinghh) [reply app/src/main/java/org/simple/clinic/home/patients/PatientsTabScreen.kt]: I have resolved this issue so this commit will be dropped now. I found out the commit where the missed changes were.

File: app/src/main/java/org/simple/clinic/util/moshi/LocalDateMoshiAdapter.kt (hunk 1 above)
Comment (janhavisinghh) [reply app/src/main/java/org/simple/clinic/util/moshi/LocalDateMoshiAdapter.kt:35]: @vinaysshenoy as far as I could understand the json qualifier annotation is used when we need different encodings for the same json type.. so for example, if there is a hex code and two integer in a json object we will use `@JsonQualifier` to differentiate the hex code in our code as colour is represented as integer type. I didn't understand how can I use this to put the condition if the date formatter is of a certain type do this and if it's not do something else.

Comment (janhavisinghh) [reply app/src/main/java/org/simple/clinic/util/moshi/LocalDateMoshiAdapter.kt:35]: Got it

File: app/src/main/java/org/simple/clinic/patient/Gender.kt (hunk 6 above)
Comment (janhavisinghh) [reply app/src/main/java/org/simple/clinic/patient/Gender.kt:74]: We don't. Removing it.

File: app/src/sharedTest/java/org/simple/clinic/TestData.kt (hunk 4 above)
Comment (janhavisinghh) [reply app/src/sharedTest/java/org/simple/clinic/TestData.kt:1231]: No. I was getting `DateTimeParseException` when I was going that as the default local date pattern is `yyyy-MM-dd`

File: app/src/test/java/org/simple/clinic/instantsearch/InstantSearchUpdateTest.kt (hunk 5 above)
Comment (janhavisinghh) [reply app/src/test/java/org/simple/clinic/instantsearch/InstantSearchUpdateTest.kt:351]: This is because we are saving the `dateOfBirth` that is being returned in the json response as `LocalDate` type. But, the format we want to prefill the `dateOfBirth` in `PatientEntryScreen` is in `dd/MM/yyyy` which is of `String` type. Which is what I'm trying to test here. If, the `LocalDate` is being converted into the correct format or not.

File: app/src/main/java/org/simple/clinic/newentry/PatientEntryEffectHandler.kt (hunk 11 above)
Comment (janhavisinghh) [reply app/src/main/java/org/simple/clinic/newentry/PatientEntryEffectHandler.kt:108]: I made this change part of this PR because when we prefilled the data then we added data to the `colonyOrVillage` field therefore making `entry.address != null` and the district and state fields to not being prefilled. Which kinda broke the previous flow.

File: app/src/main/java/org/simple/clinic/scanid/IndiaNHIDGender.kt (hunk 8 above)
Comment (janhavisinghh) [reply app/src/main/java/org/simple/clinic/scanid/IndiaNHIDGender.kt:52]: Ummm. Fair enough.

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchModel.kt]: That name exposes the implementation details. Instead, we should name this based on usage. Example: `canBePrefilled`.

File: app/src/sharedTest/java/org/simple/clinic/TestData.kt (hunk 4 above)
Comment (vinaysshenoy) [reply app/src/sharedTest/java/org/simple/clinic/TestData.kt:1231]: Yeah, this means that you are passing the date in the wrong format to `LocalDate.parse`. This version of the factory method should only be used when the date string is coming from a source that you don't have control over, which is not the case here.

File: app/src/test/java/org/simple/clinic/instantsearch/InstantSearchUpdateTest.kt (hunk 5 above)
Comment (vinaysshenoy) [reply app/src/test/java/org/simple/clinic/instantsearch/InstantSearchUpdateTest.kt:345]: Yeah, but this is the wrong place to do it. The place this needs to be tested is at the time of scanning the QR code, because once you parse a `LocalDate`, it is the same regardless of what format the string originally was,

Comment (vinaysshenoy) [reply app/src/test/java/org/simple/clinic/instantsearch/InstantSearchUpdateTest.kt:345]: As mentioned earlier, the only thing this test is doing is verifying the behaviour of `LocalDate.parse`, which is something we should not be testing.

File: app/src/main/java/org/simple/clinic/instantsearch/InstantSearchScreenKey.kt [hunk 16]
```diff
@@ -5,11 +5,13 @@ import kotlinx.android.parcel.IgnoredOnParcel
...
 import org.simple.clinic.navigation.v2.ScreenKey
 import org.simple.clinic.patient.businessid.Identifier
+import org.simple.clinic.scanid.PatientPrefillInfo
 
 @Parcelize
 data class InstantSearchScreenKey(
     val additionalIdentifier: Identifier?,
-    val initialSearchQuery: String?
+    val initialSearchQuery: String?,
+    val patientPrefillInfo: PatientPrefillInfo?
```
[the remark this answers is not in the document]
Comment (janhavisinghh) [reply app/src/main/java/org/simple/clinic/instantsearch/InstantSearchScreenKey.kt:14]: Created a [story](https://app.clubhouse.io/simpledotorg/story/3639/resolve-coupling-between-scanid-and-instantsearch-packages?vc_group_by=day&ct_workflow=all) for this will resolve in another PR.

File: app/src/main/java/org/simple/clinic/patient/OngoingNewPatientEntry.kt [hunk 17]
```diff
@@ -113,6 +114,17 @@ data class OngoingNewPatientEntry(
   fun withZone(zone: String): OngoingNewPatientEntry =
       copy(address = addressOrBlank().copy(zone = zone))
 
+  fun withPatientPrefillInfo(patientProfileInfo: PatientPrefillInfo, identifier: Identifier): OngoingNewPatientEntry =
```
[the remark this answers is not in the document]
Comment (janhavisinghh) [reply app/src/main/java/org/simple/clinic/patient/OngoingNewPatientEntry.kt]: Created a [story](https://app.clubhouse.io/simpledotorg/story/3639/resolve-coupling-between-scanid-and-instantsearch-packages?vc_group_by=day&ct_workflow=all) for this will resolve in another PR.

File: app/src/main/java/org/simple/clinic/shortcodesearchresult/ShortCodeSearchResultScreen.kt [hunk 18]
```diff
@@ -173,7 +173,11 @@ class ShortCodeSearchResultScreen :
   }
 
   override fun openPatientSearch() {
-    val screenKey = InstantSearchScreenKey(additionalIdentifier = null, initialSearchQuery = null)
+    val screenKey = if (features.isEnabled(Feature.InstantSearch)) {
+      InstantSearchScreenKey(additionalIdentifier = null, initialSearchQuery = null, patientPrefillInfo = null)
+    } else {
+      PatientSearchScreenKey(additionalIdentifier = null).wrap()
+    }
```
[the remark this answers is not in the document]
Comment (janhavisinghh) [reply app/src/main/java/org/simple/clinic/shortcodesearchresult/ShortCodeSearchResultScreen.kt]: Ok, so this is where I messed up the rebase. I am fixing this. This commit only adds patientPrefillInfo as a parameter in the openPatientSearch method

**Comment by janhavisinghh** [comment]
@vinaysshenoy I have resolved all the issues. PTAL

File: app/src/main/java/org/simple/clinic/scanid/IndiaNHIDDateOfBirthMoshiAdapter.kt [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+import com.squareup.moshi.FromJson
+import com.squareup.moshi.ToJson
+import java.time.LocalDate
+import java.time.format.DateTimeFormatter
+import java.util.Locale
+
+class IndiaNHIDDateOfBirthMoshiAdapter {
+  companion object {
+    private val dateTimeFormatter = DateTimeFormatter.ofPattern("d/M/yyyy", Locale.ENGLISH)
+  }
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/scanid/IndiaNHIDDateOfBirthMoshiAdapter.kt:12]: This should not be a static constant, can we make it a local constant please?

File: app/src/test/java/org/simple/clinic/instantsearch/InstantSearchUpdateTest.kt [hunk 20]
```diff
@@ -290,6 +301,77 @@ class InstantSearchUpdateTest {
...
+        identifier = identifier)
+
+    updateSpec
+        .given(searchQueryModel.patientPrefillInfoUpdated(patientPrefillInfo).additionalIdentifierUpdated(identifier))
+        .whenEvent(RegisterNewPatientClicked)
+        .then(assertThatNext(
+            hasNoModel(),
+            hasEffects(SaveNewOngoingPatientEntry(ongoingNewPatientEntry))
+        ))
+  }
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/instantsearch/InstantSearchUpdateTest.kt:373]: I think I may not have been clear earlier. What I mean is that this test is unnecessary since a `LocalDate` itself does not have anything to do with the format once it is parsed.

Since the actual parsing here is handled by the `ScanSimpleIdScreen`, in the event that the DOB in the scanned QR code is a different format than we expect, the parsing will fail there and we will never end up in this case since the `LocalDate` will never get created in the first place.

Hence, this test is unnecessary and we can remove it.

Does this make sense?

File: app/src/main/java/org/simple/clinic/scanid/IndiaNHIDGender.kt [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+  MALE, FEMALE, UNKNOWN;
+
+  companion object {
+    fun fromIndiaNHIDToGender(value: IndiaNHIDGender): Gender {
+      return when (value) {
+        MALE -> Gender.Male
+        FEMALE -> Gender.Female
+        else -> Gender.Unknown("U")
+      }
+    }
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/scanid/IndiaNHIDGender.kt:17]: What I meant was to add this factory function to the `Gender` type and call it `fromIndiaNHIDGender`. `IndiaNHIDGender` should not be responsible for converting itself to the `Gender` type, it should be the other way around.

**Review by vinaysshenoy — CHANGES_REQUESTED** [review]
Overall PR looks to be in good shape, left a few comments and then we'll be good to go.

**Review by vinaysshenoy — APPROVED** [review]
